src/editor/Model/Actions.re

```diff
@@ -7,7 +7,7 @@
 open Oni_Core.Types;
 open Oni_Extensions;
 
-type t =
+type t('a) =
```

Comment:
I *really* wish I knew of another way to do this, `State.t` cannot be referenced in `WindowManager.re` because it would cause a circular dependency which ocaml doesn't allow.
so we pass the split type a `'a` and make all the places that use it take a type param of `State.t` :sob:

Reply:
Hmm, ya, this is tricky! It's rough that the `'a` type sort of propagates through across a bunch of types. But I don't know a great way off the top of my head.

A couple options I was thinking of:

## Option 1: Use `() => React.syntheticElement` instead of `State.t => React.syntheticElement`

This idea essentially uses some _sleight of hand_ to use a `unit => React.syntheticElement` function as our split (while still allowing the `state` to be used!). This is nice because the definition for `WindowManager` is agnostic of the state.

I coded this up here since I thought it might be easier to see via code than to describe:
https://github.com/Akin909/oni2/pull/1/files

I think this approach might be preferable, because we can avoid the avalanche of `'a` types!

## Option 2: Recursive Types

One thing I was thinking about is if we could use recursive types in the `State.re` module, like:
```
type layout = 
| VerticalLeft
| VerticalRight
| HorizontalTop
| HorizontalBottom

type t = {
  mode: Mode.t,
  tabs: list(Tab.t),
  buffers: BufferMap.t,
...
  inputControlMode: Input.controlMode,
  iconTheme: IconTheme.t,
  languageInfo: LanguageInfo.t,
  statusBar: StatusBarModel.t,
  windows: window,
}
and windowSplit = {
    id: int,
    component: t => React.syntheticElement,
    layout,
    width: option(int),
    height: option(int),
}
and window = {
    splits: IntMap.t(windowSplit)
};
```

But the consequence of that is we'd need to move some of the other parts of `WindowManager` there too 🤔 It's kind of a trade-off in either case. But maybe including some of the `WindowManager` methods - like `createSplit`.
Not 100% sure if this works